### PR TITLE
feat: github PR comment

### LINKED ISSUE #936: Add screenshots to Github PR comment
- Add new and failures screenshots to PR commit
- Be careful of not posting several comments if there is multiple push

### PR DESCRIPTION
**Pull request by jsfez**
fix #936

### CODE REVIEW

File: apps/build-notification/src/notifications.ts
Comment (gregberge) [inline apps/build-notification/src/notifications.ts:254]: prCommentEnabled?

**Review by gregberge — COMMENTED** [review]
Good start.

- We have to design the PR comment, what do we want in it? What is really useful?

- Handle the comment removed case. Updating a deleted comment probably returns an error. We should ignore it or create another comment? Probably ignore it if deleted. Because it is intentional by a owner of the project.

File: apps/build-notification/src/notifications.ts
Comment (jsfez) [reply apps/build-notification/src/notifications.ts:254]: It's a feature switch

**Review by gregberge — APPROVED** [review]
Ok for me but I don't think you should merge it until 100% done. We have features in the project that are "pending" because they were not finished. So let this pr opened in draft and when 100% done we merge it.

**Comment by jsfez** [comment]
Comment via curl

**Comment by jsfez** [comment]
Comment via curl 2

**Comment by jsfez** [comment]
Comment via curl 3

**Comment by jsfez** [comment]
> * We have to design the PR comment, what do we want in it? What is really useful?

Great point. With the option to send either Markdown or HTML, we have many options.

Additionally, I propose including a link to turn off PR comments. Given our past experience with irritating Vercel comments, Argos users may appreciate this feature.

> * Handle the comment removed case. Updating a deleted comment probably returns an error. We should ignore it or create another comment? Probably ignore it if deleted. Because it is intentional by a owner of the project.

I've addressed the 404 error by simply ignoring it. If a user decides to delete a comment, we shouldn't add another.

**Comment by jsfez** [comment]
<p style="color:red">Red title</p>

**Comment by gregberge** [comment]
> Additionally, I propose including a link to turn off PR comments. Given our past experience with irritating Vercel comments, Argos users may appreciate this feature.

Not a link but an option in settings and a mention in the docs, definitively. It should be part of v1.

> I've addressed the 404 error by simply ignoring it. If a user decides to delete a comment, we shouldn't add another.

Agree.



> Great point. With the option to send either Markdown or HTML, we have many options.

Just be simple, the same things as the status but in bullet points, example:

---

The latest updates on your projects. Learn more about [Argos for Git ↗︎](https://argos-ci.com/docs/github-learn-more)

🔄 Waiting for screenshots

---

The latest updates on your projects. Learn more about [Argos for Git ↗︎](https://argos-ci.com/docs/github-learn-more)

**[Build #323](https://argos-ci.com/build/xxx)**

🔄 Comparing screenshots with baseline

---

The latest updates on your projects. Learn more about [Argos for Git ↗︎](https://argos-ci.com/docs/github-learn-more)

**[Build #323](https://argos-ci.com/build/xxx)**

- 2 changed (including 1 unstable)
- 28 unchanged

⚠️ Waiting for approval, [review on Argos](https://argos-ci.com/build/xxx)

---

The latest updates on your projects. Learn more about [Argos for Git ↗︎](https://argos-ci.com/docs/github-learn-more)

**[Build #323](https://argos-ci.com/build/xxx)**

- 2 changed (including 1 unstable)
- 28 unchanged

✅ Changes accepted

---

The latest updates on your projects. Learn more about [Argos for Git ↗︎](https://argos-ci.com/docs/github-learn-more)

**[Build #323](https://argos-ci.com/build/xxx)**

- 2 changed (including 1 unstable)
- 28 unchanged

❌ Changes rejected

--- 


For orphan builds and other custom types, just put the text displayed in tooltip. I think it is one of the biggest value actually.

**Comment by jsfez** [comment]
I suggest to hold off on posting comments until the Argos build is created to prevent spam.
Also, let's follow our initial plan and copy the industry leader for now.


Capture d’écran 2023-07-09 à 19 10 20

| Build name | Status | Changed | Unchanged | Updated (UTC) |
|------------|--------|---------|-----------|---------------|
| E2E | :hourglass: Waiting for approval ([inspect](#)) | 🧐 2 | ✔️ 3 | Jun 29, 2023 9:57am |
| Components | ✅  No change detected ([inspect](#)) | 🧐 0 | ✔️ 0 | Jun 29, 2023 9:57am |

**Comment by jsfez** [comment]
I see your point and agree that this feature is not good. 

However, I would like to clarify my intentions because they may not have been clear. My aim is to easily pinpoint any changes and ensure the new version aligns with the expected result. 

Adding a zoom function, could indeed enhance the user experience but doesn't serve this purpose.

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts]: You should type the return type.

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts]:
```suggestion
      return "👍 approved";
```

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts]: Should be 👎 for rejected.

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts]: "pending" and "progress" are not the same thing:

- Pending: waiting for screenshots
- Progress: diffing screenshots

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts]: throw?

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts:104]: Not the good place to aquire lock. If it blocks at this line in two jobs, two PR comments will be created. You have to do in the lock:

- Try to get existing commentId
- Create comment

The update should not be part of the lock. The lock is to get a comment id.

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts:123]:
```suggestion
      await pullRequest.$clone().$query().patch({ commentId: data.id });
```

There is a bug in `patch` it removes properties from the model, prefer to `$clone` before.

File: apps/web/src/api/v2/util.ts
Comment (gregberge) [inline apps/web/src/api/v2/util.ts:110]: Please do it outside transaction. There is no need to do this in the transaction.

Please do it in a lock to ensure there is no concurrency problem while creating the PR.

File: apps/web/src/api/v2/util.ts
Comment (gregberge) [inline apps/web/src/api/v2/util.ts:121]: If you need a cast here, there is a problem somewhere.

File: apps/database/src/models/Build.ts
Comment (gregberge) [inline apps/database/src/models/Build.ts:79]: Should be on the PR, not on the build.

**Review by gregberge — CHANGES_REQUESTED** [review]

File: apps/web/src/api/v2/util.ts
Comment (jsfez) [reply apps/web/src/api/v2/util.ts:110]: Since I test pull request existence in `getOrCreatePullRequest` function, I think that a transaction would have the same effect than a lock. Don't you agree ?

File: apps/web/src/api/v2/util.ts
Comment (jsfez) [reply apps/web/src/api/v2/util.ts:121]: I can't tell. There is number field for foreign keys that are typed has "string" in model, all over the project. I have no idea what should I do clean this mess

Comment (jsfez) [reply apps/web/src/api/v2/util.ts:121]: I'll remove the cast and go along to satisfy the linter

File: apps/build-notification/src/notifications.ts
Comment (gregberge) [reply apps/build-notification/src/notifications.ts:276]: You can set commit status and comment pr in parallel, split in several methods.

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts:92]: You should put in in a method `getOrCreateCommentId`

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts:113]: The  `if` here should not happen, we should always have a commentId.

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts:118]: You should mark the comment as deleted in the PR and not update it any more.

File: apps/build-notification/src/notifications.ts
Comment (gregberge) [inline apps/build-notification/src/notifications.ts:289]: Split the if, you don't have to fetch PR if `prCommentEnabled` is `false`.

File: apps/database/src/models/Build.ts
Comment (gregberge) [inline apps/database/src/models/Build.ts:125]: Belong to one.

File: apps/database/src/models/Project.ts
Comment (gregberge) [inline apps/database/src/models/Project.ts:35]:
```suggestion
      prCommentEnabled: { type: "boolean" },
```

File: apps/database/src/models/Project.ts
Comment (gregberge) [inline apps/database/src/models/Project.ts:46]:
```suggestion
  prCommentEnabled!: boolean;
```

File: apps/database/src/models/PullRequest.ts
Comment (gregberge) [inline apps/database/src/models/PullRequest.ts:16]: Add a `commentDeleted`.

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [inline apps/build-notification/src/githubPrComment.ts:7]: `@argos-ci/web` should not be a dependency of `built-notification`. You should create a `@argos-ci/common` or `@argos-ci/util`.

File: apps/database/migrations/20230708213033_github_pr_comment.js
Comment (gregberge) [inline apps/database/migrations/20230708213033_github_pr_comment.js:12]: Actually it is integer used by GitHub.

File: apps/database/migrations/20230708213033_github_pr_comment.js
Comment (gregberge) [inline apps/database/migrations/20230708213033_github_pr_comment.js:11]:
```suggestion
    table.integer("number").notNullable();
```

File: apps/database/migrations/20230708213033_github_pr_comment.js
Comment (gregberge) [inline apps/database/migrations/20230708213033_github_pr_comment.js:21]: Put `default(true).notNullable()`. We don't have lot of project, it can be done safely.

File: apps/database/src/models/PullRequest.ts
Comment (gregberge) [inline apps/database/src/models/PullRequest.ts:27]:
```suggestion
        relation: Model.HasManyRelation,
```

File: apps/database/src/models/PullRequest.ts
Comment (gregberge) [inline apps/database/src/models/PullRequest.ts:15]: Project should not be part of pull-request.

File: apps/web/src/api/v2/util.ts
Comment (gregberge) [inline apps/web/src/api/v2/util.ts:27]: Remove projectId from PR, a PR is linked to a github repository if you want, but not to a project.

File: apps/database/migrations/20230708213033_github_pr_comment.js
Comment (gregberge) [inline apps/database/migrations/20230708213033_github_pr_comment.js:10]: You should have a foreign key to a githubRepository, not to a project.

**Review by gregberge — CHANGES_REQUESTED** [review]

File: apps/database/migrations/20230708213033_github_pr_comment.js
Comment (gregberge) [inline apps/database/migrations/20230708213033_github_pr_comment.js:5]: Please name it `github_pull_requests`. Soon we will have GitLab.

**Review by gregberge — CHANGES_REQUESTED** [review]

File: apps/database/db/structure.sql
Comment (gregberge) [inline apps/database/db/structure.sql:178]: Please rename it to `githubPullRequestId`.

**Review by gregberge — CHANGES_REQUESTED** [review]

File: apps/build-notification/src/githubPrComment.ts
Comment (jsfez) [reply apps/build-notification/src/githubPrComment.ts:92]: I don't understand how should I use the Redis lock. I try a last attempt alone. It's I'm still out of the scope, we should have a small discussion to understand how to use it. 

For the moment, your comment is a puzzle for me.

File: apps/build-notification/src/githubPrComment.ts
Comment (jsfez) [reply apps/build-notification/src/githubPrComment.ts:113]: Now I'm sure that we are not agree

File: apps/build-notification/src/commentGithubPR.ts
Comment (gregberge) [inline apps/build-notification/src/commentGithubPR.ts:123]:
```suggestion
      if (pullRequest.commentId) {
```

**Review by gregberge — APPROVED** [review]

File: apps/build-notification/src/githubPrComment.ts
Comment (gregberge) [reply apps/build-notification/src/githubPrComment.ts]: Good for me.

**Pull request merged**
